### PR TITLE
chore(tracing): improve debug logging for v05 encoding errors

### PR DESCRIPTION
**Pull request by mabdinur**
There is an edge case were the string table used by the v0.5 encoder swaps entries. This corrupts trace payloads. In this PR I will add debug logs that will help us reproduce the behavior that triggered incident-24455.

## RISK

This PR maybe increase memory usage when v0.5 is enabled. With this change we will hold a reference to a all encoded spans until the encoded bytes are flushed and we will log trace payloads that can have a size of up to 50MB when we traces are misencoded.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: tests/integration/utils.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def send_invalid_payload_and_get_logs(encoder_cls=BadEncoder):
 def parametrize_with_all_encodings(env=None, out="", err=""):
     if env is None:
         env = dict()
-    return pytest.mark.subprocess(parametrize={"DD_TRACE_API_VERSION": ["v0.5", "v0.4"]}, env=env, out=out, err=err)
+    return pytest.mark.subprocess(parametrize={"DD_TRACE_API_VERSION": ["v0.5"]}, env=env, out=out, err=err)
```
Comment (mabdinur) [inline tests/integration/utils.py:43]:
```suggestion
    return pytest.mark.subprocess(parametrize={"DD_TRACE_API_VERSION": ["v0.5",  "v0.4"]}, env=env, out=out, err=err)
```

**Review by emmettbutler — APPROVED** [review]

File: releasenotes/notes/repro-v05-issues-cabf288f1f6cec59.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+other:
+  - |
```
Comment (ZStriker19) [inline releasenotes/notes/repro-v05-issues-cabf288f1f6cec59.yaml:3]: does this need a release note?

File: ddtrace/internal/writer/writer.py [hunk 3]
```diff
@@ -395,6 +396,15 @@ def _flush_queue_with_client(self, client, raise_exc=False):
             encoded = client.encoder.encode()
             if encoded is None:
                 return
+        except EncodingValidationError as e:
+            log.error("Encoding Error (or span was modified after finish): %s", str(e))
+            if hasattr(client.encoder, "stable"):
+                log.debug("Malformed String table: %s", client.encoder.stable.string_table_values)
+            # After detecting a malformed payload disable logging for error payloads. This feature
+            # has the potentical to generate thousands of logs each
```
Comment (mabdinur) [inline ddtrace/internal/writer/writer.py:404]:
```suggestion
            # has the potential to generate thousands of logs each containing up to 50MB of data.
```

File: releasenotes/notes/repro-v05-issues-cabf288f1f6cec59.yaml (hunk 2 above)
Comment (mabdinur) [reply releasenotes/notes/repro-v05-issues-cabf288f1f6cec59.yaml:3]: Honestly no. I added one so I could backport this change to 1.20. I will remove it.

File: tests/integration/test_integration.py [hunk 4]
```diff
@@ -531,6 +531,32 @@ def encode_traces(self, traces):
...
+    from tests.utils import AnyStringWithText
+
+    with mock.patch("ddtrace.internal.writer.writer.log") as log:
+        span = tracer.trace("name")
+        span.finish()
+        span.name = "new_name"
+        tracer.flush()
+
+    log.error.assert_has_calls([mock.call("Encoding Error (or span was modified after finish): %s", AnyStr())])
+
```
Comment (mabdinur) [inline tests/integration/test_integration.py:556]:
```suggestion
```

File: ddtrace/internal/writer/writer.py [hunk 5]
```diff
@@ -395,6 +396,15 @@ def _flush_queue_with_client(self, client, raise_exc=False):
...
             if encoded is None:
                 return
+        except EncodingValidationError as e:
+            log.error("Encoding Error (or span was modified after finish): %s", str(e))
+            if hasattr(e, "_debug_message"):
+                log.debug(e._debug_message)
+            # After detecting a malformed payload disable logging for error payloads. This feature
+            # has the potential to generate thousands of logs each containing up to 50MB of data.
+            config._trace_writer_log_err_payload = False
+            self._flush_queue_with_client(client, raise_exc)
```
Comment (mabdinur) [inline ddtrace/internal/writer/writer.py:406]:
```suggestion
```

File: tests/integration/test_integration.py (hunk 4 above)
Comment (mabdinur) [inline tests/integration/test_integration.py:556]:
```suggestion
```

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/internal/_encoding.pyx [hunk 6]
```diff
@@ -722,9 +734,16 @@ cdef class MsgpackEncoderV05(MsgpackEncoderBase):
                     PyLong_FromLong(<long> self.get_buffer()),
                     <Py_ssize_t> super(MsgpackEncoderV05, self).size,
                 )
-                return self._st.flush()
+                v05bytes = self._st.flush()
+                self._verify_encoding(v05bytes)
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/_encoding.pyx:738]: AFAICT there is no `msgpack` guaranteed to be installed here (it's not a dependency of ddtrace) which means `flush` is broken.

**Pull request merged**